src/dr_llm/project/project_info.py

```
@@ -0,0 +1,245 @@
+def _find_available_port(
+    claimed_ports: set[int],
+    base: int = BASE_PORT,
+    max_attempts: int = 100,
+) -> int:
+    for offset in range(max_attempts):
+        port = base + offset
+        if port not in claimed_ports and _port_is_free(port):
+            return port
+    raise RuntimeError(
+        f"No available port found in range {base}–{base + max_attempts - 1}"
+    )
```

Ruff (0.15.7)

[warning] Avoid specifying long messages outside the exception class

(TRY003)